dotcom-rendering/src/web/layouts/lib/enumEntries.ts

```diff
@@ -0,0 +1,39 @@
+const isEntry = (
+	tuple: [string | number, string | number],
+): tuple is [string, number] =>
+	typeof tuple[0] === 'string' && typeof tuple[1] === 'number';
+
+/**
+ * TypeScript enums have values both as keys and values, so this
+ * method filters out keys that are created from values.
+ *
+ * @example
+ * enumEntries(ArticlePillar);
+ * [
+ * 	["News", 0],
+ * 	["Opinion", 1],
+ * 	["Sport", 2],
+ * 	["Culture", 3],
+ * 	["Lifestyle",4]
+ * ]
+ *
+ * // Compare with :
+ * Object.entries(ArticlePillar);
+ * [
+ * 	["0", "News"],
+ * 	["1", "Opinion"],
+ * 	["2", "Sport"],
+ * 	["3", "Culture"],
+ * 	["4", "Lifestyle"],
+ * 	["News", 0],
+ * 	["Opinion", 1],
+ * 	["Sport", 2],
+ * 	["Culture", 3],
+ * 	["Lifestyle",4]
+ * ];
+ *
+ * @see https://www.typescriptlang.org/play?#code/KYOwrgtgBAggTgFwJYGMA2wAKS1oIZxQDeAsAFBSVQBywA7gM5QC8UADADTlVQDyADkhBIA9iBZQAjFwpUAyvxGIJAJhk8AwmDQIwcYBIDM6qgBkkAM2AMEATwwSALDIC+5cijEMRGAHRoRAHMACl4AIwArYBQEX1AEOCRrYPhkdCwcfDgASmyAbnIAegAqcgBtbioygCI2ao4oatpGaoBdE0oayXrGgSFREDaOqBqVHuqFJQQhys7qw3GtHT1gGdk5x3HzKxt7VfbZkab6Bh62A-WjvuExHskLnhrJxB6VB6rqpd19HsN3ue21jsGHqjla5FaUHIxUKQA
+ */
+export const enumEntries = <T extends Record<string | number, string | number>>(
+	record: T,
+): Array<[keyof T, number]> => Object.entries(record).filter(isEntry);
```

Comment:
I agree that it is a shame that looping over a Typescript enum gives strange results but I'm not sure it's a good idea to add more complexity to mitigate this. In all honesty, I don't really understand this code. This is a failing on my part because I don't know enough about Typescript, but I don't think I'm alone in lacking this knowledge.

The verbose if statements inside the original for loops are indeed messy and far from pretty but they are the simplest way to solve this problem and don't require people to learn new abstractions or understand special Typescript concepts.

Maybe we improve this code by adding more comments explaining why there is a strange if statement checking for `isNan`?

Reply:
Yes, an alternative would be to simplify the types altogether to be string unions in `@guardian/libs`, and stop using the extremely confusing no-quite-enums of TypeScript and stick with idiomatic JavaScript.

```ts
// the options array exported if you want to loop over the options
const options = ["a", "b", "c"] as const;

// the type which as no runtime overhead at all
type Option = typeof options[number]; // "a" | "b" | "c"

// the predicate for filtering or other runtime checks
const isOption = (option: string): option is Option => options.include(option)


export { options, isOption }
export type { Option }
```

Reply:
[Captured in this gist](https://gist.github.com/mxdvl/2e5fa0562b35a4fc9d18dcd5ac791b0f)

Reply:
To be honest as a testing util I don't think abstracting some of this complexity away is the worst thing in the world!

Most of the time when you're writing on and working on tests, you don't want to get bogged down in how to deal with typescripts awkward enum implementations, so if you can get a tidy and easy to use abstraction which is verbose and clear in what it outputs then I don't think its a big deal